bach/bach/operations/value_propagation.py

```diff
@@ -0,0 +1,138 @@
+"""
+Copyright 2022 Objectiv B.V.
+"""
+from enum import Enum
+from typing import Optional, Union, Sequence, List, Dict
+from bach import DataFrame
+
+
+class ValuePropagationMethod(Enum):
+    BACK_FILL = 'bfill'
+    FORWARD_FILL = 'ffill'
+
+    @classmethod
+    def get_method(cls, method: str) -> 'ValuePropagationMethod':
+        if method in ('ffill', 'pad'):
+            return cls.FORWARD_FILL
+
+        if method in ('bfill', 'backfill'):
+            return cls.BACK_FILL
+
+        raise ValueError(f'"{method}" is not a valid propagation method.')
+
+
+class ValuePropagation:
+    """
+    In order to instantiate this class you should provide the following params:
+    df: a DataFrame containing all series to be filled by value propagation
+    method: a ValuePropagationMethod to be used.
+        - FORWARD_FILL: Fill missing values by propagating the last non-nullable value in each series.
+        - BACK_FILL: Fill missing values by using the next non-nullable value in each series.
+    returns a new DataFrame with filled missing values.
+
+    Steps of how values are propagated:
+    1.  Number each row of the sorted DataFrame. The main reason of numbering each row, is because
+        bfill is the opposite to ffill, so we can say that it uses a reversed sorting. The issue with this
+        is that Bach will always sort NULLS LAST, therefore reversing the order by won't work. By numbering
+        each row, we can at least rely on a non-nullable value.
+
+    2. Sort dataframe based on the row number.
+            Ascending if method == 'ffill' and descending if method == 'bfill'.
+    3. Partition each series to be filled, a partition is the cumulative sum of the total amount
+        of observed non-nullable values in the series.
+    4. Group the series to be filled by its partition and assign the first value.
+    """
+    ROW_NUMBER_SERIES_NAME = '_row_number'
```

Comment:
```suggestion
    ROW_NUMBER_SERIES_NAME = '__row_number'
```

We use double underscores for reserved column names in other places (e.g. indexing.py and describe.py)

Reply:
Done here https://github.com/objectiv/objectiv-analytics/pull/831/commits/0aed7652b20199fb3c2d1132c2870c856f9c8514